Code/Editor/Viewport.cpp

```diff
@@ -1154,6 +1137,29 @@ bool QtViewport::HitTest(const QPoint& point, HitContext& hitInfo)
     return false;
 }
 
+AZ::Vector3 QtViewport::GetHitLocation(const QPoint& point)
+{
+    Vec3 pos = Vec3(ZERO);
+    HitContext hit;
+    if (HitTest(point, hit))
+    {
+        pos = hit.raySrc + hit.rayDir * hit.dist;
+        pos = SnapToGrid(pos);
+    }
+    else
+    {
+        bool hitTerrain;
+        pos = ViewToWorld(point, &hitTerrain);
+        if (hitTerrain)
+        {
+            pos.z = GetIEditor()->GetTerrainElevation(pos.x, pos.y);
+        }
+        pos = SnapToGrid(pos);
+    }
+
+    return AZ::Vector3(pos.x, pos.y, pos.z);
+}
```

Comment:
I think it's great this is fixed but one thing I'm a bit concerned about is we're still relying on a lot of legacy functionality here. Ideally we'd not be relying on `GetIEditor->GetTerrainElevation` (what is that even returning? Is it just coincidence it returns the position where the ground mesh is now? If you move that mesh does this now fail?). I think we really need a task to kill a lot more of this legacy functionality and use the new AZ Viewport work (we have `ViewportScreen` for `ScreenToWorld/WorldToScreen` etc.. as well) and use intersection queries against atom meshes, potentially using the InstersectorBus that exists (we use this in the new camera). Hopefully it's something we can pick-up in the coming months.

Reply:
I totally agree, to be clear though in this case the "else" case isn't actually being used. There was a bug similar to this awhile back for drag/drop assets into the viewport where I rewrote the QtViewport::HitTest method to indeed use the new viewport interaction queries, so all I did for this was move the existing code to a new method so I could use it as well for the right-click create entity. The fallback else case that is calling GetTerrainElevation isn't being hit in the gif I show above, that was just there from legacy. I left it alone for now because there might indeed be something that is relying on it, but I agree we should do a pass sooner rather than later and get rid of the legacy logic still lurking around.

Reply:
Okay perfect thanks for clarifying - I'd say it'd be good to just delete the `GetIEditor()->GetTerrainElevation` as even if it's being called, if we get the right answer it's by chance 😛 Totally agree would be great to do another pass to remove some of this legacy code. Cheers!

Reply:
So... just a bit of color around GetTerrainElevation.  I'm totally great with it being removed since it's completely a legacy remnant, but FWIW that function _will_ actually work as-is with the new terrain system going in.  Back in LY 1.x, we started the refactor for the new terrain system, so we put the new terrain bus in place as a part of extracting the legacy terrain system into a gem.  The GetTerrainElevation method calls that new terrain bus, so with the new terrain system we'll build, if you have a terrain in your level, it will get the actual height of the terrain at that point.

Reply:
Haha okay nice 🙂 Good to know! (though I dream of a day when `GetIEditor` no longer exists 😜)